pydicom/fileset.py

```diff
@@ -0,0 +1,2742 @@
+            addition to - or removal from - the File-set.
+        """
+        if not path and self.path is None:
+            raise ValueError(
+                "The path to the root directory is required for a "
+                "new File-set"
+            )
+
+        if path and self.path:
+            raise ValueError(
+                "The path for an existing File-set cannot be changed, use "
+                "'FileSet.copy()' to write the File-set to a new location"
+            )
+
+        if path:
+            self._path = Path(path)
+
+        # Don't write unless changed or new
+        if not self.is_staged:
+            return
+
+        # Path to the DICOMDIR file
+        p = self._path / 'DICOMDIR'
+
+        # Write just the DICOMDIR if:
+        #   * only the File-set Idenfitication module elements have changed, or
+        #   * only moves are required and `dicomdir_only` is True
+        major_change = bool(self._stage['+']) or bool(self._stage['-'])
+        if dicomdir_only and major_change:
+            raise RuntimeError(
+                "'Fileset.write()' called with 'dicomdir_only' but changes to "
+                "the File-set's managed instances are staged"
+            )
+
+        if not dicomdir_only:
+            major_change |= bool(self._stage['~'])
+
+        if dicomdir_only and not major_change:
+            with open(p, 'wb') as f:
+                f = DicomFileLike(f)
+                self._write_dicomdir(f, force_implicit=force_implicit)
+
+            self._ds = dcmread(p)
+            self._stage['^'] = False
+            return
+
+        # Worst case scenario if all instances in one directory
+        if len(self) > 10**6:
+            self._use_alphanumeric = True
+        if len(self) > 36**6:
+            raise NotImplementedError(
+                "pydicom doesn't support writing File-sets with more than "
+                "2176782336 managed instances"
+            )
+
+        # Remove the removals - must be first because the File IDs will be
+        #   incorrect with the removals still in the tree
+        for instance in self._stage['-'].values():
+            try:
+                Path(instance.path).unlink()
+            except FileNotFoundError:
+                pass
+            self._tree.remove(instance.node)
+
+        # We need to be careful not to overwrite the source file
+        #   for a different (later) instance
+        # Check for collisions between the new and old File IDs
+        #   and move any to the stage
+        fout = {Path(ii.FileID) for ii in self}
+        fin = {
+            ii.node._file_id for ii in self
+            if ii.SOPInstanceUID not in self._stage['+']
+        }
+        collisions = fout & fin
+        for instance in [ii for ii in self if ii.node._file_id in collisions]:
+            self._stage['+'][instance.SOPInstanceUID] = instance
+            instance._flags.move = False
+            instance._apply_stage('+')
+            shutil.move(
```

Comment:
Change to `copy` in case we get an error during `write()` so we don't end up with data loss. Check that we don't end up with loose files though.

Reply:
Actually, since all these are collisions we won't end with with loose files.